aip/0180.md

```diff
@@ -0,0 +1,126 @@
+   to communicate correctly with a newer server. In other words, not only are
+   inputs and outputs compatible, but the serialization and deserialization
+   expectations continue to match.
+3. Semantic compatibility: Code written against a previous version **must**
+   continue to receive the intended and expected results.
+
+### Adding components
+
+In general, new components (interfaces, methods, messages, fields, enums, or
+enum values) **may** be added to existing APIs in the same major version.
+
+However, keep the following guidelines in mind when doing this:
+
+- Code written against the previous surface (and thus is unaware of the new
+  components) **must** continue to be treated the same way as before.
+  - New required fields **must not** be added to existing request messages or
+    resources.
+  - Any field being populated by clients **must** have a default behavior
+    matching the behavior before the field was introduced.
+  - Any field previously populated by the server **must** continue to be
+    populated, even if it introduces redundancy.
+- Enums **may** be freely added to request messages. For response messages and
+  resources, be aware that it is possible that user code does not handle new
+  values gracefully. In this situation, enums which are expected to receive new
+  values **should** document this, and while enum values still **may** be added
+  in this situation, appropriate caution **should** be used.
+
+**Note:** It is possible when adding a component closely related to an existing
+component (for example, `string foo_value` when `string foo` already exists) to
+enter a situation where generated code will conflict. Service owners **should**
+be aware of subtleties in the tooling they or their users are likely to use
+(and tool authors **should** endeavor to avoid such subtleties if possible).
+
+### Removing or renaming components
+
+Existing components (interfaces, methods, messages, fields, enums, or enum
+values) **must not** be removed from existing APIs in the same major version.
+Removing a component is a backwards incompatible change.
+
+**Important:** Renaming a component is semantically equivalent to "remove and
+add". In cases where these sorts of changes are desirable, a service **may**
+add the new component, but **must not** remove the existing one.
+
+### Changing the type of fields
+
+Existing fields and messages **must not** have their type changed, even if the
+new type is wire-compatible, because type changes alter generated code in a
+breaking way.
+
+**Note:** An exception to this rule is that `int32` values **may** be changed
+to `int64`.
+
+### Changing resource names
+
+A resource **must not** change its name.
+
+Unlike most breaking changes, this affects major versions as well: in order for
+a client to expect to use v2.0 access a resource that was created in v1.0 or
+vice versa, the same resource name **must** be used in both versions.
+
```

Comment:
It may be worth talking about the IAM sort of issue here - if a resource changes in v2 in a way that could cause problems for v1 customers, it's okay to fail v1 requests that ask for resources containing those aspects. (Not a good description, but I hope you know what I mean.)

Basically, interoperability of the same resource across major versions is thorny, and should be considered carefully, but it's okay for it not to reach "perfection". We should probably have a whole section about this - but we need to talk about it first!

Reply:
It's definitely worth avoiding where possible, but I believe there are cases where that means an API is simply at a dead-end, and can't implement much-desired new features, ever.
Put it this way - if we don't talk about it, API producers are still likely to do it, but without specific guidance :)

Reply:
In that case, I would argue that this should be a separate AIP that specifically discusses this question. It seems like a pretty nuanced issue and this AIP is already on the long side. Would that be reasonable? (If so, can you open an issue?)

Reply:
Fine by me, and will open an issue now.